# Protease Activity Analysis (PAA)
Analyzing the activity of proteases and their substrates is critical to defining the
biological functions of these enzymes and to designing new diagnostics and therapeutics
that target protease dysregulation in disease.

To facilitate protease research, we present Protease Activity Analysis (PAA). PAA is a software package with a collection of tools for analyzing protease activity data in Python. PAA leverages an object-oriented implementation to provide a modular framework for streamlined analysis across three major components.  

1. PAA provides a facile framework to query datasets of synthetic peptide substrates and their cleavage susceptibilities across a diverse set of proteases. 
2. PAA includes tools for the automated analysis and visualization of user-input enzyme-substrate activity measurements generated through in vitro screens against synthetic peptide substrates. 
3. PAA can be used to analyze in vivo protease activity signatures that are generated by activity-based sensors using a set of modular machine learning functions.


## Setup
1. Install the following dependencies: [conda 3.x](https://docs.conda.io/en/latest/miniconda.html)
2. [Generate Github SSH keys](https://help.github.com/en/github/authenticating-to-github/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent). If you already have SSH keys you can first [check to make sure](https://help.github.com/en/github/authenticating-to-github/checking-for-existing-ssh-keys).
3. [Add your SSH key to your Github account](https://help.github.com/en/github/authenticating-to-github/adding-a-new-ssh-key-to-your-github-account).
4. Download this repository: 
```
git clone git@github.com:apsoleimany/protease_activity_analysis.git
```


## Installation
To use the protease activity analysis (paa) toolbox, first create the environment and then install the package:
```
cd protease_activity_analysis
conda env create -f environment.yml
conda activate paa
pip install -e .
```

## Usage
To enter the protease environment: `conda activate paa`

Once inside the environment, the package can be directly imported and used in a Python shell/script: 
```
>>> import protease_activity_analysis as paa
>>> paa.tests.test_installation()
```

To deactivate after you are done: `conda deactivate`

## Tutorials
To get started using PAA, we recommend you select the tutorial in the 'Notebooks' folder that is most aligned with your application of interest. 

### Database.py Tutorial
This tutorial describes the functionality of the 'database' subpackage of the PAA repo. The subpackage incorporates the SubstrateDatabase class that provides a framework to load and query datasets of protease-substrate activity.

To demo the package, we show how to load and query a rich dataset of 150 unique peptide substrates and their cleavage susceptibility against a diverse set of recombinant proteases, generated in the Bhatia Lab. All input files required query said datasets are made publically available and should comprise a valuable research for the sceintific community.

The SubstrateDatabase class enables to:
- Query protease-substrate activity assay data for proteases, substrates, or sequences of itnerest
- Identification of the substrates present in the database that are most robustly and specifically cleaved by a protease of interest
- Identification of the proteases that robustly and/or specifically cleave a peptide substrate of interest
- Assessment of the degree of similarity between a sequence of interest and sequences in the database
- Assessment of degree of similarity between all substrates in the database
- Search for k-mer motifs across all sequences in the database

![database.pdf](https://github.com/apsoleimany/protease_activity_analysis/files/8173160/database.pdf)


### Tutorial KineticDataset
This tutorial describes the functionality of the 'kinetic' subpackage of the PAA repo. The subpackage incorporates the KineticDataset class that provides a framework to load and analyze raw data files from multiplexed fluorogenic assays generated directly by measurement instruments (e.g., fluorimeters). Additionally, the package supports visualization of inputs from retrospective screens, for which a matrix summarizing cleavage efficiencies across a set of samples may have already been produced.

To demo the package, we show how to load raw files from a retrospective in vitro screen of a panel of lung cancer-associated recombinant proteases against a panel of 14 peptide substrate from the Bhatia Lab (Kirkpatrick et al., STM, 2020). 

In this tutorial you will learn how to: 
- Plot and analyze data from each independent fluorogenic run.
- Aggregate summary data outputs form individual runs to generate a matrix of samples of interest vs. peptides of interest.
- Plot aggregated data using relevant visualization tools.
![kinetic.pdf](https://github.com/apsoleimany/protease_activity_analysis/files/8173193/kinetic.pdf)


### Substrate.py and Protease.py tutorial
This tutorial describes the functionality of the 'substrate' and 'protease' subpackages of the PAA repo.

The 'substrate' package is intended to help analyze and visualize peptide sequences.
The 'protease' package has functions related to classifying proteases by class and finding protease orthologs across species.
To demo the package, we show how to create a new instance of the SubstarteDatabase class in 'database.py' and test the different functions on all the substrate sequences in the database.
![substrate_protease.pdf](https://github.com/apsoleimany/protease_activity_analysis/files/8173248/substrate_protease.pdf)


## Contributing
Code should be written in compliance with the [Google Python style guide](http://google.github.io/styleguide/pyguide.html) with [Google style docstrings](http://queirozf.com/entries/python-docstrings-reference-examples#google-style).
